{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Laboratórios de Bioinformática\n",
    "### Análise da sequência e das features presentes no NCBI\n",
    "\n",
    "Grupo 6: Joana Gabriel, Maria Couto, Teresa Coimbra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_idlist(termo_de_pesquisa):\n",
    "    '''Recebe uma string que será o termo de pesquisa\n",
    "    Retorna o conjunto de ids dos records que resultam dessa pesquisa (uma lista)'''\n",
    "    from Bio import Entrez\n",
    "    Entrez.email = 'lbiogrupo6@gmail.com'\n",
    "    handle = Entrez.esearch(db='protein', term = termo_de_pesquisa, retmax = 100) #pesquisar na base de dados\n",
    "    record = Entrez.read(handle)\n",
    "    handle.close()\n",
    "    idlist = record['IdList']\n",
    "    return idlist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ids(str1,str2, idlist):\n",
    "    '''Recebe duas strings p.ex: str1 = nome do gene e do organismo, str2 = idlist\n",
    "    Retorna uma lista de IDS filtrados de acordo com a pesquisa'''\n",
    "    from Bio import Entrez\n",
    "    res = []\n",
    "    for id1 in idlist: \n",
    "        handle = Entrez.efetch(db='protein', id = id1 , rettype=\"gb\", retmode=\"text\")\n",
    "        Entrez.email = 'lbiogrupo6@gmail.com'\n",
    "        record = handle.read()\n",
    "        if (str1 in record) and (str2 in record):\n",
    "            res.append(id1)\n",
    "    \n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_seqio(res_ids):\n",
    "    '''Recebe a lista de IDS após a pesquisa\n",
    "    Retorna uma lista de records SeqIO'''\n",
    "    from Bio import Entrez\n",
    "    from Bio import SeqIO\n",
    "    RES = []\n",
    "    for id0 in res_ids:\n",
    "        Entrez.email = 'lbiogrupo6@gmail.com'\n",
    "        handle = Entrez.efetch(db=\"protein\", id=id0, rettype=\"gb\", retmode=\"text\")\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "        RES.append(record)\n",
    "    return RES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fasta(ids):\n",
    "    '''Recebe uma lista de ids para um dos genes\n",
    "    Cria ficheiros fasta correspondentes a cada um dos ids'''\n",
    "    from Bio import Entrez\n",
    "    from Bio import SeqIO\n",
    "    import re\n",
    "    for id0 in ids:\n",
    "        Entrez.email = 'lbiogrupo6@gmail.com'\n",
    "        handle = Entrez.efetch(db=\"protein\", id=id0, rettype=\"fasta\", retmode=\"text\")\n",
    "        record = SeqIO.read(handle, \"fasta\")\n",
    "        nome = str(record.id +'.faa')\n",
    "        nome1 = re.sub('[*,\"\\/\\\\\\[\\]:;\\|]', '', nome)  # Retirar os caracteres não permitidos para guardar o ficheiro\n",
    "        SeqIO.write(record, nome1, 'fasta')\n",
    "        \n",
    "def get_annotations(record):\n",
    "    '''Retorna um dicionario com id do record: anotação do record'''\n",
    "    a = {}  # dicionario com as annotations\n",
    "    for r in record: #para cada record\n",
    "        a[r.id] = r.annotations\n",
    "    return a      \n",
    "\n",
    "def get_features(record):\n",
    "    '''Recebe os records\n",
    "    Faz print do ID, features e qualifiers'''\n",
    "    for feat in record:\n",
    "    \n",
    "        print('\\n')\n",
    "        print('\\n')\n",
    "        print('-->ID:', feat.id)\n",
    "        n = feat.features\n",
    "        for i in range(len(record)):\n",
    "            print('*********************')\n",
    "            print('Features: ')\n",
    "            print(n[i])                \n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'NP_112420.1': {'topology': 'linear', 'data_file_division': 'PRI', 'date': '10-JAN-2021', 'accessions': ['NP_112420'], 'sequence_version': 1, 'db_source': 'REFSEQ: accession NM_031157.4', 'keywords': ['RefSeq', 'MANE Select'], 'source': 'Homo sapiens (human)', 'organism': 'Homo sapiens', 'taxonomy': ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo'], 'references': [Reference(title='The nuclear localization sequence mediates hnRNPA1 amyloid fibril formation revealed by cryoEM structure', ...), Reference(title='Post-translational modifications of hnRNP A1 differentially modulate retroviral IRES-mediated translation initiation', ...), Reference(title='Heterogeneous Nuclear Ribonucleoprotein A1 (hnRNP A1) and hnRNP A2 Inhibit Splicing to Human Papillomavirus 16 Splice Site SA409 through a UAG-Containing Sequence in the E7 Coding Region', ...), Reference(title='A nuclear localization domain in the hnRNP A1 protein', ...), Reference(title='Assignment of the human heterogeneous nuclear ribonucleoprotein A1 gene (HNRPA1) to chromosome 12q13.1 by cDNA competitive in situ hybridization', ...), Reference(title='Modeling by homology of RNA binding domain in A1 hnRNP protein', ...), Reference(title='Alternative splicing in the human gene for the core protein A1 generates another hnRNP protein', ...), Reference(title='Isolation of an active gene encoding human hnRNP protein A1. Evidence for alternative splicing', ...), Reference(title='cDNA cloning of human hnRNP protein A1 reveals the existence of multiple mRNA isoforms', ...), Reference(title='Mammalian single-stranded DNA binding protein UP I is derived from the hnRNP core protein A1', ...)], 'comment': 'REVIEWED REFSEQ: This record has been curated by NCBI staff in\\ncollaboration with Francesco Amaldi. The reference sequence was\\nderived from CD705336.1, BC070315.1, BU541100.1, BC052296.1 and\\nAC078778.34.\\nThis sequence is a reference standard in the RefSeqGene project.\\nSummary: This gene encodes a member of a family of ubiquitously\\nexpressed heterogeneous nuclear ribonucleoproteins (hnRNPs), which\\nare RNA-binding proteins that associate with pre-mRNAs in the\\nnucleus and influence pre-mRNA processing, as well as other aspects\\nof mRNA metabolism and transport. The protein encoded by this gene\\nis one of the most abundant core proteins of hnRNP complexes and\\nplays a key role in the regulation of alternative splicing.\\nMutations in this gene have been observed in individuals with\\namyotrophic lateral sclerosis 20. Multiple alternatively spliced\\ntranscript variants have been found. There are numerous pseudogenes\\nof this gene distributed throughout the genome. [provided by\\nRefSeq, Feb 2016].\\nTranscript Variant: This variant (2, also known as A1B) contains an\\nalternate in-frame exon compared to variant 1, resulting in a\\nlonger protein (isoform b) than isoform 1.\\nPublication Note:  This RefSeq record includes a subset of the\\npublications that are available for this gene. Please see the Gene\\nrecord to access additional publications.', 'structured_comment': OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'SRR1803617.274967.1, SRR1803614.26916.1 [ECO:0000332]'), ('RNAseq introns', 'single sample supports all introns SAMEA1965299, SAMEA1966682 [ECO:0000348]')])), ('RefSeq-Attributes', OrderedDict([('MANE Ensembl match', 'ENST00000340913.11/ ENSP00000341826.7'), ('RefSeq Select criteria', 'based on manual assertion, conservation, longest protein')]))]), 'molecule_type': 'protein'}, 'NP_002127.1': {'topology': 'linear', 'data_file_division': 'PRI', 'date': '10-JAN-2021', 'accessions': ['NP_002127'], 'sequence_version': 1, 'db_source': 'REFSEQ: accession NM_002136.4', 'keywords': ['RefSeq'], 'source': 'Homo sapiens (human)', 'organism': 'Homo sapiens', 'taxonomy': ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo'], 'references': [Reference(title='The nuclear localization sequence mediates hnRNPA1 amyloid fibril formation revealed by cryoEM structure', ...), Reference(title='Post-translational modifications of hnRNP A1 differentially modulate retroviral IRES-mediated translation initiation', ...), Reference(title='Heterogeneous Nuclear Ribonucleoprotein A1 (hnRNP A1) and hnRNP A2 Inhibit Splicing to Human Papillomavirus 16 Splice Site SA409 through a UAG-Containing Sequence in the E7 Coding Region', ...), Reference(title='A nuclear localization domain in the hnRNP A1 protein', ...), Reference(title='Assignment of the human heterogeneous nuclear ribonucleoprotein A1 gene (HNRPA1) to chromosome 12q13.1 by cDNA competitive in situ hybridization', ...), Reference(title='Modeling by homology of RNA binding domain in A1 hnRNP protein', ...), Reference(title='Alternative splicing in the human gene for the core protein A1 generates another hnRNP protein', ...), Reference(title='Isolation of an active gene encoding human hnRNP protein A1. Evidence for alternative splicing', ...), Reference(title='cDNA cloning of human hnRNP protein A1 reveals the existence of multiple mRNA isoforms', ...), Reference(title='Mammalian single-stranded DNA binding protein UP I is derived from the hnRNP core protein A1', ...)], 'comment': 'REVIEWED REFSEQ: This record has been curated by NCBI staff in\\ncollaboration with Francesco Amaldi. The reference sequence was\\nderived from CD705336.1, BC052296.1 and AC078778.34.\\nSummary: This gene encodes a member of a family of ubiquitously\\nexpressed heterogeneous nuclear ribonucleoproteins (hnRNPs), which\\nare RNA-binding proteins that associate with pre-mRNAs in the\\nnucleus and influence pre-mRNA processing, as well as other aspects\\nof mRNA metabolism and transport. The protein encoded by this gene\\nis one of the most abundant core proteins of hnRNP complexes and\\nplays a key role in the regulation of alternative splicing.\\nMutations in this gene have been observed in individuals with\\namyotrophic lateral sclerosis 20. Multiple alternatively spliced\\ntranscript variants have been found. There are numerous pseudogenes\\nof this gene distributed throughout the genome. [provided by\\nRefSeq, Feb 2016].\\nTranscript Variant: This variant (1, also known as A1) encodes\\nisoform a.\\nPublication Note:  This RefSeq record includes a subset of the\\npublications that are available for this gene. Please see the Gene\\nrecord to access additional publications.', 'structured_comment': OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'SRR1803612.84113.1, SRR1660809.65599.1 [ECO:0000332]'), ('RNAseq introns', 'single sample supports all introns SAMEA1965299, SAMEA1966682 [ECO:0000348]')]))]), 'molecule_type': 'protein'}, 'NP_001011724.1': {'topology': 'linear', 'data_file_division': 'PRI', 'date': '15-DEC-2020', 'accessions': ['NP_001011724'], 'sequence_version': 1, 'db_source': 'REFSEQ: accession NM_001011724.3', 'keywords': ['RefSeq'], 'source': 'Homo sapiens (human)', 'organism': 'Homo sapiens', 'taxonomy': ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo'], 'references': [Reference(title='Phylogenetic-based propagation of functional annotations within the Gene Ontology consortium', ...)], 'comment': 'VALIDATED REFSEQ: This record has undergone validation or\\npreliminary review. The reference sequence was derived from\\nBX477483.1, AK094423.1, CD523364.1 and AL137058.8.\\nTranscript Variant: This variant (1) represents the longer\\ntranscript. Variants 1 and 2 encode the same protein.', 'structured_comment': OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'SRR1803612.172647.1, SRR1660807.263050.1 [ECO:0000332]')]))]), 'molecule_type': 'protein'}, 'NP_001011725.1': {'topology': 'linear', 'data_file_division': 'PRI', 'date': '14-DEC-2020', 'accessions': ['NP_001011725', 'XP_495955'], 'sequence_version': 1, 'db_source': 'REFSEQ: accession NM_001011725.3', 'keywords': ['RefSeq'], 'source': 'Homo sapiens (human)', 'organism': 'Homo sapiens', 'taxonomy': ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo'], 'references': [Reference(title='Phylogenetic-based propagation of functional annotations within the Gene Ontology consortium', ...)], 'comment': \"VALIDATED REFSEQ: This record has undergone validation or\\npreliminary review. The reference sequence was derived from\\nBX477483.1, BQ425201.1 and AL137058.8.\\nOn Feb 9, 2005 this sequence version replaced XP_495955.1.\\nTranscript Variant: This variant (2) differs in the 5' UTR compared\\nto variant 1. Variants 1 and 2 encode the same protein.\", 'structured_comment': OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'BC082242.1, SRR3476690.936319.1 [ECO:0000332]')]))]), 'molecule_type': 'protein'}}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-->ID: NP_112420.1\n",
      "*********************\n",
      "Features: \n",
      "type: source\n",
      "location: [0:372]\n",
      "qualifiers:\n",
      "    Key: chromosome, Value: ['12']\n",
      "    Key: db_xref, Value: ['taxon:9606']\n",
      "    Key: map, Value: ['12q13.13']\n",
      "    Key: organism, Value: ['Homo sapiens']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Protein\n",
      "location: [0:372]\n",
      "qualifiers:\n",
      "    Key: calculated_mol_wt, Value: ['38616']\n",
      "    Key: note, Value: ['heterogeneous nuclear ribonucleoprotein B2 protein; nuclear ribonucleoprotein particle A1 protein; helix-destabilizing protein; single-strand DNA-binding protein UP1; heterogeneous nuclear ribonucleoprotein core protein A1; heterogeneous nuclear ribonucleoprotein A1B protein; hnRNP core protein A1-like 3; single-strand RNA-binding protein; putative heterogeneous nuclear ribonucleoprotein A1-like 3; epididymis secretory sperm binding protein']\n",
      "    Key: product, Value: ['heterogeneous nuclear ribonucleoprotein A1 isoform b']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Site\n",
      "location: [0:1]\n",
      "qualifiers:\n",
      "    Key: note, Value: ['N-acetylmethionine. /evidence=ECO:0000244|PubMed:22814378; propagated from UniProtKB/Swiss-Prot (P09651.5)']\n",
      "    Key: site_type, Value: ['acetylation']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Site\n",
      "location: [1:2]\n",
      "qualifiers:\n",
      "    Key: note, Value: ['N-acetylserine, in Heterogeneous nuclear ribonucleoprotein A1, N-terminally processed. /evidence=ECO:0000244|PubMed:20068231, ECO:0000244|PubMed:22814378, ECO:0000269|Ref.9; propagated from UniProtKB/Swiss-Prot (P09651.5)']\n",
      "    Key: site_type, Value: ['acetylation']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-->ID: NP_002127.1\n",
      "*********************\n",
      "Features: \n",
      "type: source\n",
      "location: [0:320]\n",
      "qualifiers:\n",
      "    Key: chromosome, Value: ['12']\n",
      "    Key: db_xref, Value: ['taxon:9606']\n",
      "    Key: map, Value: ['12q13.13']\n",
      "    Key: organism, Value: ['Homo sapiens']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Protein\n",
      "location: [0:320]\n",
      "qualifiers:\n",
      "    Key: calculated_mol_wt, Value: ['34065']\n",
      "    Key: note, Value: ['heterogeneous nuclear ribonucleoprotein B2 protein; nuclear ribonucleoprotein particle A1 protein; helix-destabilizing protein; single-strand DNA-binding protein UP1; heterogeneous nuclear ribonucleoprotein core protein A1; heterogeneous nuclear ribonucleoprotein A1B protein; hnRNP core protein A1-like 3; single-strand RNA-binding protein; putative heterogeneous nuclear ribonucleoprotein A1-like 3; epididymis secretory sperm binding protein']\n",
      "    Key: product, Value: ['heterogeneous nuclear ribonucleoprotein A1 isoform a']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Region\n",
      "location: [11:92]\n",
      "qualifiers:\n",
      "    Key: db_xref, Value: ['CDD:241205']\n",
      "    Key: note, Value: ['RNA recognition motif 1 in heterogeneous nuclear ribonucleoprotein A1 (hnRNP A1) and similar proteins; cd12761']\n",
      "    Key: region_name, Value: ['RRM1_hnRNPA1']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Site\n",
      "location: order{[11:12], [14:15], [16:17], [18:20], [41:42], [43:44], [45:46], [54:57], [58:59], [81:82], [84:85], [86:92]}\n",
      "qualifiers:\n",
      "    Key: db_xref, Value: ['CDD:241205']\n",
      "    Key: note, Value: ['DNA binding site [nucleotide binding]']\n",
      "    Key: site_type, Value: ['DNA binding']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-->ID: NP_001011724.1\n",
      "*********************\n",
      "Features: \n",
      "type: source\n",
      "location: [0:320]\n",
      "qualifiers:\n",
      "    Key: chromosome, Value: ['13']\n",
      "    Key: db_xref, Value: ['taxon:9606']\n",
      "    Key: map, Value: ['13q14.3']\n",
      "    Key: organism, Value: ['Homo sapiens']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Protein\n",
      "location: [0:320]\n",
      "qualifiers:\n",
      "    Key: calculated_mol_wt, Value: ['34094']\n",
      "    Key: note, Value: ['hnRNP A1-like 2; hnRNP core protein A1-like 2']\n",
      "    Key: product, Value: ['heterogeneous nuclear ribonucleoprotein A1-like 2']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Region\n",
      "location: [3:94]\n",
      "qualifiers:\n",
      "    Key: note, Value: ['propagated from UniProtKB/Swiss-Prot (Q32P51.2)']\n",
      "    Key: region_name, Value: ['Globular A domain']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Site\n",
      "location: [5:6]\n",
      "qualifiers:\n",
      "    Key: note, Value: ['Phosphoserine. /evidence=ECO:0000250|UniProtKB:P49312; propagated from UniProtKB/Swiss-Prot (Q32P51.2)']\n",
      "    Key: site_type, Value: ['phosphorylation']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-->ID: NP_001011725.1\n",
      "*********************\n",
      "Features: \n",
      "type: source\n",
      "location: [0:320]\n",
      "qualifiers:\n",
      "    Key: chromosome, Value: ['13']\n",
      "    Key: db_xref, Value: ['taxon:9606']\n",
      "    Key: map, Value: ['13q14.3']\n",
      "    Key: organism, Value: ['Homo sapiens']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Protein\n",
      "location: [0:320]\n",
      "qualifiers:\n",
      "    Key: calculated_mol_wt, Value: ['34094']\n",
      "    Key: note, Value: ['hnRNP A1-like 2; hnRNP core protein A1-like 2']\n",
      "    Key: product, Value: ['heterogeneous nuclear ribonucleoprotein A1-like 2']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Region\n",
      "location: [3:94]\n",
      "qualifiers:\n",
      "    Key: note, Value: ['propagated from UniProtKB/Swiss-Prot (Q32P51.2)']\n",
      "    Key: region_name, Value: ['Globular A domain']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Site\n",
      "location: [5:6]\n",
      "qualifiers:\n",
      "    Key: note, Value: ['Phosphoserine. /evidence=ECO:0000250|UniProtKB:P49312; propagated from UniProtKB/Swiss-Prot (Q32P51.2)']\n",
      "    Key: site_type, Value: ['phosphorylation']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def teste1():\n",
    "    str1 = 'heterogeneous nuclear ribonucleoprotein A1'\n",
    "    str2 = 'Homo sapiens'\n",
    "    \n",
    "    termo_de_pesquisa = 'hnrnpa1'\n",
    "    idList = get_idlist(termo_de_pesquisa)\n",
    "    \n",
    "    ids = get_ids(str1,str2,idList)\n",
    "    \n",
    "    get_fasta(ids)\n",
    "    \n",
    "    record = get_seqio(ids) #pega no resultado filtrado dos ids\n",
    "    anota = get_annotations(record)\n",
    "    print(anota)\n",
    "    feature = get_features(record)\n",
    "    if feature != None:\n",
    "        print(feature)\n",
    "        \n",
    "teste1()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'P0DTC9.1': {'topology': 'linear', 'data_file_division': 'VRL', 'date': '02-DEC-2020', 'accessions': ['P0DTC9'], 'sequence_version': 1, 'db_source': 'UniProtKB: locus NCAP_SARS2, accession P0DTC9; class: standard. created: Apr 22, 2020. sequence updated: Apr 22, 2020. annotation updated: Dec 2, 2020. xrefs: MN908947.3, QHD43423.2, YP_009724397.2, 6M3M_A, 6M3M_B, 6M3M_C, 6M3M_D, 6VYO_A, 6VYO_B, 6VYO_C, 6VYO_D, 6WJI_A, 6WJI_B, 6WJI_C, 6WJI_D, 6WJI_E, 6WJI_F, 6WKP_A, 6WKP_B, 6WKP_C, 6WKP_D, 6WZO_A, 6WZO_B, 6WZO_C, 6WZO_D, 6WZQ_A, 6WZQ_B, 6WZQ_C, 6WZQ_D, 6YI3_A, 6YUN_A, 6YUN_B, 6ZCO_A, 7C22_A, 7C22_B, 7C22_C, 7C22_D, 7CDZ_A, 7CDZ_B, 7CDZ_C, 7CDZ_D, 7CE0_A, 7CE0_B, 7CE0_C, 7CE0_D xrefs (non-sequence databases): PDBsum:6M3M, PDBsum:6VYO, PDBsum:6WJI, PDBsum:6WKP, PDBsum:6WZO, PDBsum:6WZQ, PDBsum:6YI3, PDBsum:6YUN, PDBsum:6ZCO, PDBsum:7C22, PDBsum:7CDZ, PDBsum:7CE0, BMRB:P0DTC9, SMR:P0DTC9, BioGRID:4383847, ComplexPortal:CPX-5686, IntAct:P0DTC9, GeneID:43740575, SIGNOR:P0DTC9, Proteomes:UP000464024, GO:0044172, GO:0044177, GO:0044220, GO:0019013, GO:0042802, GO:0003723, HAMAP:MF_04096, InterPro:IPR001218, InterPro:IPR037179, InterPro:IPR037195, Pfam:PF00937, PIRSF:PIRSF003888, SUPFAM:SSF103068, SUPFAM:SSF110304, PROSITE:PS51929, PROSITE:PS51928', 'keywords': ['3D-structure', 'ADP-ribosylation', 'Host Golgi apparatus', 'Phosphoprotein', 'Reference proteome', 'Ribonucleoprotein', 'RNA-binding', 'Transcription', 'Transcription regulation', 'Viral nucleoprotein', 'Virion'], 'source': 'Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2)', 'organism': 'Severe acute respiratory syndrome coronavirus 2', 'taxonomy': ['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus'], 'references': [Reference(title='A new coronavirus associated with human respiratory disease in China', ...)], 'comment': '[FUNCTION] Packages the positive strand viral genome RNA into a\\nhelical ribonucleocapsid (RNP) and plays a fundamental role during\\nvirion assembly through its interactions with the viral genome and\\nmembrane protein M. Plays an important role in enhancing the\\nefficiency of subgenomic viral RNA transcription as well as viral\\nreplication (By similarity). May modulate transforming growth\\nfactor-beta signaling by binding host smad3 (By similarity).\\n{ECO:0000250|UniProtKB:P59595, ECO:0000255|HAMAP-Rule:MF_04096}.\\n[SUBUNIT] Homooligomer. Both monomeric and oligomeric forms\\ninteract with RNA. Interacts with protein M. Interacts with protein\\nE. May bind to host HNRNPA1. Interacts with NSP3; this interaction\\nserves to tether the genome to the newly translated\\nreplicase-transcriptase complex at a very early stage of infection\\n(By similarity). May interact with host SMAD3 (By similarity).\\n{ECO:0000250|UniProtKB:P59595, ECO:0000255|HAMAP-Rule:MF_04096}.\\n[INTERACTION] P0DTC9; P0DTC9: N; NbExp=7; IntAct=EBI-25475856,\\nEBI-25475856.\\n[SUBCELLULAR LOCATION] Virion {ECO:0000255|HAMAP-Rule:MF_04096}.\\nHost endoplasmic reticulum-Golgi intermediate compartment\\n{ECO:0000255|HAMAP-Rule:MF_04096}. Host Golgi apparatus\\n{ECO:0000255|HAMAP-Rule:MF_04096}. Note=Located inside the virion,\\ncomplexed with the viral RNA. Probably associates with ER-derived\\nmembranes where it participates in viral RNA synthesis and virus\\nbudding. {ECO:0000255|HAMAP-Rule:MF_04096}.\\n[PTM] ADP-ribosylated. The ADP-ribosylation is retained in the\\nvirion during infection. {ECO:0000255|HAMAP-Rule:MF_04096}.\\n[PTM] Phosphorylated on serine and threonine residues.\\n{ECO:0000255|HAMAP-Rule:MF_04096}.\\n[SIMILARITY] Belongs to the betacoronavirus nucleocapsid protein\\nfamily. {ECO:0000255|HAMAP-Rule:MF_04096}.', 'molecule_type': 'protein'}}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-->ID: P0DTC9.1\n",
      "*********************\n",
      "Features: \n",
      "type: source\n",
      "location: [0:419]\n",
      "qualifiers:\n",
      "    Key: db_xref, Value: ['taxon:2697049']\n",
      "    Key: host, Value: ['Homo sapiens (Human)']\n",
      "    Key: organism, Value: ['Severe acute respiratory syndrome coronavirus 2']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def teste2():\n",
    "    '''Pesquisa pela nucleoproteína'''\n",
    "    str1 = 'Nucleoprotein'\n",
    "    str2 = 'SARS-CoV-2'\n",
    "    termo_de_pesquisa = 'NCAP_SARS2'\n",
    "    idlist = get_idlist(termo_de_pesquisa)\n",
    "    \n",
    "    ids = get_ids(str1,str2,idlist)\n",
    "    get_fasta(ids)\n",
    "    \n",
    "    record = get_seqio(ids) \n",
    "    anota = get_annotations(record)\n",
    "    \n",
    "    print(anota)\n",
    "    \n",
    "    feature = get_features(record)\n",
    "\n",
    "teste2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'NP_001138576.1': {'topology': 'linear', 'data_file_division': 'PRI', 'date': '03-JAN-2021', 'accessions': ['NP_001138576'], 'sequence_version': 1, 'db_source': 'REFSEQ: accession NM_001145104.2', 'keywords': ['RefSeq'], 'source': 'Homo sapiens (human)', 'organism': 'Homo sapiens', 'taxonomy': ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo'], 'references': [Reference(title='Role of AP-2alpha/TGF-beta1/Smad3 axis in rats with intervertebral disc degeneration', ...), Reference(title='Interactome Mapping Provides a Network of Neurodegenerative Disease Proteins and Uncovers Widespread Protein Aggregation in Affected Brains', ...), Reference(title='In Vitro Lineage-Specific Differentiation of Vascular Smooth Muscle Cells in Response to SMAD3 Deficiency: Implications for SMAD3-Related Thoracic Aortic Aneurysm', ...), Reference(title='Statin inhibits large hepatitis delta antigen-Smad3 -twist-mediated epithelial-to-mesenchymal transition and hepatitis D virus secretion', ...), Reference(title='Receptor-associated Mad homologues synergize as effectors of the TGF-beta response', ...), Reference(title='Mad-related genes in the human', ...), Reference(title='E2F-4, a new member of the E2F gene family, has oncogenic activity and associates with p107 in vivo', ...), Reference(title='E2F-4, a new member of the E2F transcription factor family, interacts with p107', ...), Reference(title='Loeys-Dietz Syndrome', ...), Reference(title='Heritable Thoracic Aortic Disease Overview', ...)], 'comment': \"REVIEWED REFSEQ: This record has been curated by NCBI staff. The\\nreference sequence was derived from AC012568.7.\\nSummary: The SMAD family of proteins are a group of intracellular\\nsignal transducer proteins similar to the gene products of the\\nDrosophila gene 'mothers against decapentaplegic' (Mad) and the C.\\nelegans gene Sma. The SMAD3 protein functions in the transforming\\ngrowth factor-beta signaling pathway, and transmits signals from\\nthe cell surface to the nucleus, regulating gene activity and cell\\nproliferation. It also functions as a tumor suppressor. Mutations\\nin this gene are associated with aneurysms-osteoarthritis syndrome\\nand Loeys-Dietz Syndrome 3. [provided by RefSeq, Nov 2019].\\nTranscript Variant: This variant (4) differs in the 5' UTR and\\ncoding sequence compared to variant 1. The resulting isoform (4) is\\nshorter at the N-terminus compared to isoform 1.\\nSequence Note: This RefSeq record was created from transcript and\\ngenomic sequence data to make the sequence consistent with the\\nreference genome assembly. The genomic coordinates used for the\\ntranscript record were based on transcript alignments.\\nPublication Note:  This RefSeq record includes a subset of the\\npublications that are available for this gene. Please see the Gene\\nrecord to access additional publications.\", 'structured_comment': OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'AK300614.1 [ECO:0000332]')]))]), 'molecule_type': 'protein'}, 'NP_001138575.1': {'topology': 'linear', 'data_file_division': 'PRI', 'date': '03-JAN-2021', 'accessions': ['NP_001138575'], 'sequence_version': 1, 'db_source': 'REFSEQ: accession NM_001145103.2', 'keywords': ['RefSeq'], 'source': 'Homo sapiens (human)', 'organism': 'Homo sapiens', 'taxonomy': ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo'], 'references': [Reference(title='Role of AP-2alpha/TGF-beta1/Smad3 axis in rats with intervertebral disc degeneration', ...), Reference(title='Interactome Mapping Provides a Network of Neurodegenerative Disease Proteins and Uncovers Widespread Protein Aggregation in Affected Brains', ...), Reference(title='In Vitro Lineage-Specific Differentiation of Vascular Smooth Muscle Cells in Response to SMAD3 Deficiency: Implications for SMAD3-Related Thoracic Aortic Aneurysm', ...), Reference(title='Statin inhibits large hepatitis delta antigen-Smad3 -twist-mediated epithelial-to-mesenchymal transition and hepatitis D virus secretion', ...), Reference(title='Receptor-associated Mad homologues synergize as effectors of the TGF-beta response', ...), Reference(title='Mad-related genes in the human', ...), Reference(title='E2F-4, a new member of the E2F gene family, has oncogenic activity and associates with p107 in vivo', ...), Reference(title='E2F-4, a new member of the E2F transcription factor family, interacts with p107', ...), Reference(title='Loeys-Dietz Syndrome', ...), Reference(title='Heritable Thoracic Aortic Disease Overview', ...)], 'comment': \"REVIEWED REFSEQ: This record has been curated by NCBI staff. The\\nreference sequence was derived from AC012568.7.\\nSummary: The SMAD family of proteins are a group of intracellular\\nsignal transducer proteins similar to the gene products of the\\nDrosophila gene 'mothers against decapentaplegic' (Mad) and the C.\\nelegans gene Sma. The SMAD3 protein functions in the transforming\\ngrowth factor-beta signaling pathway, and transmits signals from\\nthe cell surface to the nucleus, regulating gene activity and cell\\nproliferation. It also functions as a tumor suppressor. Mutations\\nin this gene are associated with aneurysms-osteoarthritis syndrome\\nand Loeys-Dietz Syndrome 3. [provided by RefSeq, Nov 2019].\\nTranscript Variant: This variant (3) differs in the 5' UTR and\\ncoding sequence compared to variant 1. The resulting isoform (3)\\nhas a shorter and distinct N-terminus compared to isoform 1.\\nSequence Note: This RefSeq record was created from transcript and\\ngenomic sequence data to make the sequence consistent with the\\nreference genome assembly. The genomic coordinates used for the\\ntranscript record were based on transcript alignments.\\nPublication Note:  This RefSeq record includes a subset of the\\npublications that are available for this gene. Please see the Gene\\nrecord to access additional publications.\", 'structured_comment': OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'AK298139.1 [ECO:0000332]')]))]), 'molecule_type': 'protein'}, 'NP_001138574.1': {'topology': 'linear', 'data_file_division': 'PRI', 'date': '03-JAN-2021', 'accessions': ['NP_001138574'], 'sequence_version': 1, 'db_source': 'REFSEQ: accession NM_001145102.2', 'keywords': ['RefSeq'], 'source': 'Homo sapiens (human)', 'organism': 'Homo sapiens', 'taxonomy': ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo'], 'references': [Reference(title='Role of AP-2alpha/TGF-beta1/Smad3 axis in rats with intervertebral disc degeneration', ...), Reference(title='Interactome Mapping Provides a Network of Neurodegenerative Disease Proteins and Uncovers Widespread Protein Aggregation in Affected Brains', ...), Reference(title='In Vitro Lineage-Specific Differentiation of Vascular Smooth Muscle Cells in Response to SMAD3 Deficiency: Implications for SMAD3-Related Thoracic Aortic Aneurysm', ...), Reference(title='Statin inhibits large hepatitis delta antigen-Smad3 -twist-mediated epithelial-to-mesenchymal transition and hepatitis D virus secretion', ...), Reference(title='Receptor-associated Mad homologues synergize as effectors of the TGF-beta response', ...), Reference(title='Mad-related genes in the human', ...), Reference(title='E2F-4, a new member of the E2F gene family, has oncogenic activity and associates with p107 in vivo', ...), Reference(title='E2F-4, a new member of the E2F transcription factor family, interacts with p107', ...), Reference(title='Loeys-Dietz Syndrome', ...), Reference(title='Heritable Thoracic Aortic Disease Overview', ...)], 'comment': \"REVIEWED REFSEQ: This record has been curated by NCBI staff. The\\nreference sequence was derived from AC012568.7.\\nSummary: The SMAD family of proteins are a group of intracellular\\nsignal transducer proteins similar to the gene products of the\\nDrosophila gene 'mothers against decapentaplegic' (Mad) and the C.\\nelegans gene Sma. The SMAD3 protein functions in the transforming\\ngrowth factor-beta signaling pathway, and transmits signals from\\nthe cell surface to the nucleus, regulating gene activity and cell\\nproliferation. It also functions as a tumor suppressor. Mutations\\nin this gene are associated with aneurysms-osteoarthritis syndrome\\nand Loeys-Dietz Syndrome 3. [provided by RefSeq, Nov 2019].\\nTranscript Variant: This variant (2) differs in the 5' UTR and\\ncoding sequence compared to variant 1. The resulting isoform (2) is\\nshorter at the N-terminus compared to isoform 1.\\nSequence Note: This RefSeq record was created from transcript and\\ngenomic sequence data to make the sequence consistent with the\\nreference genome assembly. The genomic coordinates used for the\\ntranscript record were based on transcript alignments.\\nPublication Note:  This RefSeq record includes a subset of the\\npublications that are available for this gene. Please see the Gene\\nrecord to access additional publications.\", 'structured_comment': OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'SRR7346977.2963354.1, AK316017.1 [ECO:0000332]')]))]), 'molecule_type': 'protein'}, 'NP_005893.1': {'topology': 'linear', 'data_file_division': 'PRI', 'date': '03-JAN-2021', 'accessions': ['NP_005893', 'NP_056215'], 'sequence_version': 1, 'db_source': 'REFSEQ: accession NM_005902.4', 'keywords': ['RefSeq', 'MANE Select'], 'source': 'Homo sapiens (human)', 'organism': 'Homo sapiens', 'taxonomy': ['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo'], 'references': [Reference(title='Role of AP-2alpha/TGF-beta1/Smad3 axis in rats with intervertebral disc degeneration', ...), Reference(title='Interactome Mapping Provides a Network of Neurodegenerative Disease Proteins and Uncovers Widespread Protein Aggregation in Affected Brains', ...), Reference(title='In Vitro Lineage-Specific Differentiation of Vascular Smooth Muscle Cells in Response to SMAD3 Deficiency: Implications for SMAD3-Related Thoracic Aortic Aneurysm', ...), Reference(title='Statin inhibits large hepatitis delta antigen-Smad3 -twist-mediated epithelial-to-mesenchymal transition and hepatitis D virus secretion', ...), Reference(title='Receptor-associated Mad homologues synergize as effectors of the TGF-beta response', ...), Reference(title='Mad-related genes in the human', ...), Reference(title='E2F-4, a new member of the E2F gene family, has oncogenic activity and associates with p107 in vivo', ...), Reference(title='E2F-4, a new member of the E2F transcription factor family, interacts with p107', ...), Reference(title='Loeys-Dietz Syndrome', ...), Reference(title='Heritable Thoracic Aortic Disease Overview', ...)], 'comment': \"REVIEWED REFSEQ: This record has been curated by NCBI staff. The\\nreference sequence was derived from AC087482.9 and AC012568.7.\\nOn Sep 20, 2004 this sequence version replaced NP_056215.1.\\nSummary: The SMAD family of proteins are a group of intracellular\\nsignal transducer proteins similar to the gene products of the\\nDrosophila gene 'mothers against decapentaplegic' (Mad) and the C.\\nelegans gene Sma. The SMAD3 protein functions in the transforming\\ngrowth factor-beta signaling pathway, and transmits signals from\\nthe cell surface to the nucleus, regulating gene activity and cell\\nproliferation. It also functions as a tumor suppressor. Mutations\\nin this gene are associated with aneurysms-osteoarthritis syndrome\\nand Loeys-Dietz Syndrome 3. [provided by RefSeq, Nov 2019].\\nTranscript Variant: This variant (1) represents the longest\\ntranscript and encodes the longest isoform (1).\\nPublication Note:  This RefSeq record includes a subset of the\\npublications that are available for this gene. Please see the Gene\\nrecord to access additional publications.\", 'structured_comment': OrderedDict([('Evidence-Data', OrderedDict([('Transcript exon combination', 'SRR1163655.67621.1, BC050743.1 [ECO:0000332]'), ('RNAseq introns', 'single sample supports all introns SAMEA1965299, SAMEA1966682 [ECO:0000348]')])), ('RefSeq-Attributes', OrderedDict([('MANE Ensembl match', 'ENST00000327367.9/ ENSP00000332973.4'), ('RefSeq Select criteria', 'based on conservation, expression, longest protein')]))]), 'molecule_type': 'protein'}}\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-->ID: NP_001138576.1\n",
      "*********************\n",
      "Features: \n",
      "type: source\n",
      "location: [0:230]\n",
      "qualifiers:\n",
      "    Key: chromosome, Value: ['15']\n",
      "    Key: db_xref, Value: ['taxon:9606']\n",
      "    Key: map, Value: ['15q22.33']\n",
      "    Key: organism, Value: ['Homo sapiens']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Protein\n",
      "location: [0:230]\n",
      "qualifiers:\n",
      "    Key: calculated_mol_wt, Value: ['25591']\n",
      "    Key: note, Value: ['hMAD-3; hSMAD3; mad homolog JV15-2; mad protein homolog; MAD, mothers against decapentaplegic homolog 3; SMAD, mothers against DPP homolog 3; SMA- and MAD-related protein 3; mad3; MAD homolog 3']\n",
      "    Key: product, Value: ['mothers against decapentaplegic homolog 3 isoform 4']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Region\n",
      "location: [28:219]\n",
      "qualifiers:\n",
      "    Key: db_xref, Value: ['CDD:199826']\n",
      "    Key: note, Value: ['C-terminal Mad Homology 2 (MH2) domain in SMAD2 and SMAD3; cd10985']\n",
      "    Key: region_name, Value: ['MH2_SMAD_2_3']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Site\n",
      "location: order{[45:46], [50:52], [53:54], [64:69], [72:73], [76:77], [79:80], [81:83], [88:90], [91:94], [96:97], [165:166], [168:169], [171:172], [175:176], [177:178], [181:184], [188:189], [190:191], [194:195], [199:200], [201:202], [203:204], [205:206], [208:210], [211:213], [215:217]}\n",
      "qualifiers:\n",
      "    Key: db_xref, Value: ['CDD:199826']\n",
      "    Key: note, Value: ['trimer interface [polypeptide binding]']\n",
      "    Key: site_type, Value: ['other']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-->ID: NP_001138575.1\n",
      "*********************\n",
      "Features: \n",
      "type: source\n",
      "location: [0:381]\n",
      "qualifiers:\n",
      "    Key: chromosome, Value: ['15']\n",
      "    Key: db_xref, Value: ['taxon:9606']\n",
      "    Key: map, Value: ['15q22.33']\n",
      "    Key: organism, Value: ['Homo sapiens']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Protein\n",
      "location: [0:381]\n",
      "qualifiers:\n",
      "    Key: calculated_mol_wt, Value: ['43106']\n",
      "    Key: note, Value: ['hMAD-3; hSMAD3; mad homolog JV15-2; mad protein homolog; MAD, mothers against decapentaplegic homolog 3; SMAD, mothers against DPP homolog 3; SMA- and MAD-related protein 3; mad3; MAD homolog 3']\n",
      "    Key: product, Value: ['mothers against decapentaplegic homolog 3 isoform 3']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Region\n",
      "location: [<24:88]\n",
      "qualifiers:\n",
      "    Key: db_xref, Value: ['CDD:260161']\n",
      "    Key: note, Value: ['N-terminal Mad Homology 1 (MH1) domain; cl00055']\n",
      "    Key: region_name, Value: ['MH1']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Region\n",
      "location: [179:370]\n",
      "qualifiers:\n",
      "    Key: db_xref, Value: ['CDD:199826']\n",
      "    Key: note, Value: ['C-terminal Mad Homology 2 (MH2) domain in SMAD2 and SMAD3; cd10985']\n",
      "    Key: region_name, Value: ['MH2_SMAD_2_3']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-->ID: NP_001138574.1\n",
      "*********************\n",
      "Features: \n",
      "type: source\n",
      "location: [0:320]\n",
      "qualifiers:\n",
      "    Key: chromosome, Value: ['15']\n",
      "    Key: db_xref, Value: ['taxon:9606']\n",
      "    Key: map, Value: ['15q22.33']\n",
      "    Key: organism, Value: ['Homo sapiens']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Protein\n",
      "location: [0:320]\n",
      "qualifiers:\n",
      "    Key: calculated_mol_wt, Value: ['35764']\n",
      "    Key: note, Value: ['hMAD-3; hSMAD3; mad homolog JV15-2; mad protein homolog; MAD, mothers against decapentaplegic homolog 3; SMAD, mothers against DPP homolog 3; SMA- and MAD-related protein 3; mad3; MAD homolog 3']\n",
      "    Key: product, Value: ['mothers against decapentaplegic homolog 3 isoform 2']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Region\n",
      "location: [<0:27]\n",
      "qualifiers:\n",
      "    Key: db_xref, Value: ['CDD:381778']\n",
      "    Key: note, Value: ['N-terminal Mad Homology 1 (MH1) domain; cl00055']\n",
      "    Key: region_name, Value: ['MH1']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Region\n",
      "location: [118:309]\n",
      "qualifiers:\n",
      "    Key: db_xref, Value: ['CDD:199826']\n",
      "    Key: note, Value: ['C-terminal Mad Homology 2 (MH2) domain in SMAD2 and SMAD3; cd10985']\n",
      "    Key: region_name, Value: ['MH2_SMAD_2_3']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "-->ID: NP_005893.1\n",
      "*********************\n",
      "Features: \n",
      "type: source\n",
      "location: [0:425]\n",
      "qualifiers:\n",
      "    Key: chromosome, Value: ['15']\n",
      "    Key: db_xref, Value: ['taxon:9606']\n",
      "    Key: map, Value: ['15q22.33']\n",
      "    Key: organism, Value: ['Homo sapiens']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Protein\n",
      "location: [0:425]\n",
      "qualifiers:\n",
      "    Key: calculated_mol_wt, Value: ['47950']\n",
      "    Key: note, Value: ['hMAD-3; hSMAD3; mad homolog JV15-2; mad protein homolog; MAD, mothers against decapentaplegic homolog 3; SMAD, mothers against DPP homolog 3; SMA- and MAD-related protein 3; mad3; MAD homolog 3']\n",
      "    Key: product, Value: ['mothers against decapentaplegic homolog 3 isoform 1']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Site\n",
      "location: [1:2]\n",
      "qualifiers:\n",
      "    Key: note, Value: ['N-acetylserine. /evidence=ECO:0000244|PubMed:19413330, ECO:0000244|PubMed:22814378; propagated from UniProtKB/Swiss-Prot (P84022.1)']\n",
      "    Key: site_type, Value: ['acetylation']\n",
      "\n",
      "*********************\n",
      "Features: \n",
      "type: Region\n",
      "location: [7:132]\n",
      "qualifiers:\n",
      "    Key: db_xref, Value: ['CDD:199815']\n",
      "    Key: note, Value: ['N-terminal Mad Homology 1 (MH1) domain in SMAD2 and SMAD3; cd10491']\n",
      "    Key: region_name, Value: ['MH1_SMAD_2_3']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def teste3():\n",
    "    '''Pesquisa pela proteína SMAD3'''\n",
    "    str1 = 'mothers against decapentaplegic homolog 3'\n",
    "    str2 = 'Homo sapiens'\n",
    "    termo_de_pesquisa = 'SMAD3'\n",
    " \n",
    "    idlist = get_idlist(termo_de_pesquisa)\n",
    "    \n",
    "    ids = get_ids(str1,str2,idlist)\n",
    "    get_fasta(ids)\n",
    "    \n",
    "    record = get_seqio(ids) \n",
    "    anota = get_annotations(record)\n",
    "    \n",
    "    print(anota)\n",
    "    \n",
    "    feature = get_features(record)\n",
    "    if feature != None:\n",
    "        print(feature)\n",
    "teste3()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
