{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Não existem variantes repetidos nas duas amostras.\n"
     ]
    }
   ],
   "source": [
    "import vcf\n",
    "\n",
    "def vcf2dict(loc):\n",
    "    \"\"\"\n",
    "    For a vcf file returns dict with all its records.\n",
    "\n",
    "    The Key:Value pair is structured as {'position':'full vcf record object'}\n",
    "    :param loc:string\n",
    "    :return:dict{int:obj}\n",
    "    \"\"\"\n",
    "    reader = vcf.Reader(open(loc, 'r'))\n",
    "    records = {}\n",
    "    for record in reader:\n",
    "        records[record.POS] = record\n",
    "    return records\n",
    "\n",
    "def vcf_isec(vcf1, vcf2):\n",
    "    \"\"\"\n",
    "    For 2 vcf dicts structured as {'position':'full vcf record object'}, returns the intersection of variants in a similarly structured dict.\n",
    "    :param vcf1: dict{int:obj}\n",
    "    :param vcf2: dict{int:obj}\n",
    "    :return: dict{int:obj}\n",
    "    \"\"\"\n",
    "    isec_records = {}\n",
    "    for record in vcf1.keys():\n",
    "        #https://stackoverflow.com/questions/3733992/determine-whether-a-key-is-present-in-a-dictionary\n",
    "        if record in vcf2:\n",
    "            isec_records[record]=vcf2[record]\n",
    "    # Logic could be added to check if dict is empty or not as in:\n",
    "    # https://stackoverflow.com/questions/23177439/python-checking-if-a-dictionary-is-empty-doesnt-seem-to-work\n",
    "    # But for modularity this list will just return and empty dict\n",
    "    return isec_records\n",
    "\n",
    "# Import the Portuguese (ERR4157959) and Spanish (ERR4395294) sample\n",
    "pt_records = vcf2dict('D:/Bioinformatica/Laboratorios/trabalho/variantes/pt.raw.vcf')\n",
    "es_records = vcf2dict('D:/Bioinformatica/Laboratorios/trabalho/variantes/es.raw.vcf')\n",
    "\n",
    "\n",
    "# Find intersection\n",
    "pt_es_isec = vcf_isec(pt_records,es_records)\n",
    "\n",
    "# Produce human readable answer (checks if there is an isec or not)\n",
    "if pt_es_isec:\n",
    "    print(\"Existem variantes presentes nas duas amostras, estes estão localizados em {0}\".format(pt_es_isec.keys()))\n",
    "else:\n",
    "    print(\"Não existem variantes repetidos nas duas amostras.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Recorremos ao seguinte site para saber o id do ensembl do nosso gene de interesse (N):\n",
    "https://covid-19.ensembl.org/Sars_cov_2/Gene/Summary?db=core;g=ENSSASG00005000005;r=MN908947.3:28274-29533;t=ENSSAST00005000005\n",
    "\n",
    "Os resultados foram filtrados para apresentar apenas o gene de interesse: ENSSASG00005000005"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The spanish sample has detected variations in the positions:\n",
      "[28657, 28863] for the following studied gene ENSSASG00005000005\n",
      "\n",
      "The portuguese sample has detected variations in the positions:\n",
      "[28881, 28882, 28883] for the following studied gene ENSSASG00005000005\n"
     ]
    }
   ],
   "source": [
    "from BCBio import GFF\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqFeature\n",
    "from collections import defaultdict\n",
    "\n",
    "\n",
    "def gff2rec(in_file):\n",
    "    \"\"\"\n",
    "    For a gff or gtf file returns list with all its genes.\n",
    "    :param in_file:string\n",
    "    :return:list[object]\n",
    "    \"\"\"\n",
    "    in_handle = open(in_file)\n",
    "    # https://biopython.org/wiki/GFF_Parsing\n",
    "    # This method is not good for massive gff files, read the doc\n",
    "    records = []\n",
    "    for record in GFF.parse(in_handle):\n",
    "        # This is dirty Python\n",
    "        records.append(record)\n",
    "    in_handle.close()\n",
    "    return records\n",
    "\n",
    "def seq_rec_sgenes(record):\n",
    "    \"\"\"\n",
    "    For SeqRecord returns list with all its genes that might have been studied for function.\n",
    "    :param record:string\n",
    "    :return:list[object]\n",
    "    \"\"\"\n",
    "    studied_genes = []\n",
    "    for feature in record.features:\n",
    "        if feature.type == \"gene\" and feature.qualifiers['gene_name'] and feature.qualifiers['gene_biotype']:\n",
    "            studied_genes.append(feature)\n",
    "    return studied_genes\n",
    "\n",
    "def lookup_var2genes(variations, record):\n",
    "    \"\"\"\n",
    "    For a list of SeqFeature objects, it performs a lookup of variations provided in a dict structured as {'position':'pyvcf record object'}.\n",
    "\n",
    "    This operation returns a dict where for gene there is a list of variation positions.\n",
    "    :param variations: dict{int:obj}\n",
    "    :param record: list[object]\n",
    "    :return: dict{string:list[int]}\n",
    "    \"\"\"\n",
    "    genes_var = defaultdict(list)\n",
    "    for gene in seq_rec_sgenes(record):\n",
    "        # The \"gene_id\" qualifier is a list of len 1 so I have to call the index 0 otherwise in futher opperations I will try to add a list as a dict key.\n",
    "        # This raises an error as list are not hashable\n",
    "        gene_id = gene.qualifiers['gene_id'][0]\n",
    "        genes_var[gene_id] = []\n",
    "        for var in variations.keys():\n",
    "            # Variant position\n",
    "            # Gene location using the logic in:\n",
    "            # https://biopython.readthedocs.io/en/latest/api/Bio.SeqFeature.html#Bio.SeqFeature.FeatureLocation\n",
    "            if gene.location.start < variations[var].POS < gene.location.end:\n",
    "                # If the var is within the start and end of the gene, it means it belongs to it\n",
    "                # how to add to lists inside dict:\n",
    "                # https://stackoverflow.com/questions/26367812/appending-to-list-in-python-dictionary\n",
    "                genes_var[gene_id].append(var)\n",
    "    return genes_var\n",
    "\n",
    "# Import SarsCov2 ref seq genes\n",
    "SarsCov2_file = \"D:/Bioinformatica/Laboratorios/trabalho/variantes/Sars_cov_2.ASM985889v3.101.gtf\"\n",
    "sars_cov_rec = gff2rec(SarsCov2_file)[0]\n",
    "\n",
    "\n",
    "genes_with_var_es = lookup_var2genes(es_records, sars_cov_rec)\n",
    "genes_with_var_pt = lookup_var2genes(pt_records, sars_cov_rec)\n",
    "\n",
    "\n",
    "# Print the gene variations per gene\n",
    "for gene in genes_with_var_es:\n",
    "    if genes_with_var_es[gene]:\n",
    "        if gene == 'ENSSASG00005000005':\n",
    "            # Dont print gene with no\n",
    "            print(\"The spanish sample has detected variations in the positions:\\n{0} for the following studied gene {1}\\n\".format(genes_with_var_es[gene], gene))\n",
    "for gene in genes_with_var_pt:\n",
    "    if genes_with_var_pt[gene]:\n",
    "        if gene == 'ENSSASG00005000005':\n",
    "            # Dont print gene with no\n",
    "            print(\"The portuguese sample has detected variations in the positions:\\n{0} for the following studied gene {1}\".format(genes_with_var_pt[gene], gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
